### PR TITLE
Add organiser

### PR DESCRIPTION
**Pull request by pg20**

### CODE REVIEW

File: app/controllers/organisers_controller.rb [hunk 1]
```diff
@@ -4,6 +4,21 @@ class OrganisersController < ApplicationController
   def index
     @organisers = User.where(role: 1)
   end
+  
+
+  def create
+    @user = User.find_by(email: params[:email])
+    if @user.blank? || (@user.role != "organizer")
```
Comment (DMips) [inline app/controllers/organisers_controller.rb]: you're not saving an user here. why? it's supposed to be 'create' action ;) Also why you're checking the role? It doesn't matter whether user with "abc@abc.com" exists but his role is "mentee". Creating organizer in such case should be impossible as based on the role user will have access to different page. Unless we want people to be able to be both organizer and mentee but than how will we decide where to redirect them and what kind of links to show them? I'm for using one email address per role.

File: app/mailers/organisers_mailer.rb [hunk 2]
```diff
@@ -0,0 +1,8 @@
+class OrganisersMailer < ApplicationMailer
+   
+    def organisers_add_email(user)
+      @user = user
+      @url = edit_organiser_url(SecureRandom.hex(8))
```
Comment (DMips) [inline app/mailers/organisers_mailer.rb]: you should create this token while saving organizer and assign it to this user (token - new field in table user)

File: app/views/organisers/create.html.slim [hunk 3]
```diff
@@ -0,0 +1 @@
+aa
```
Comment (DMips) [inline app/views/organisers/create.html.slim]: please delete it as we don't want to use this view

File: app/views/organisers_mailer/organisers_add_email.text.slim [hunk 4]
```diff
@@ -0,0 +1,5 @@
+| Hello,
```
Comment (DMips) [inline app/views/organisers_mailer/organisers_add_email.text.slim]: please change `text.slim` to `slim.html` and use html tags here - look in other mailers for reference :)

File: config/routes.rb [hunk 5]
```diff
@@ -2,8 +2,7 @@
 
 
   root 'home#index'
-
-  resources :organisers, only: [:index]
+  resources :organisers #only: [:index]
```
Comment (DMips) [inline config/routes.rb]: you can use `resources :organisers only: [:index, :create]`

File: app/views/organisers/index.html.slim [hunk 6]
```diff
@@ -5,6 +5,13 @@
       p
       p
       p
+      = form_tag organisers_path, :method => :post do
```
Comment (DMips) [inline app/views/organisers/index.html.slim]: don't forget to display errors when user couldn't be saved as email was empty/ invalid/ already taken

File: test/mailers/organisers_mailer_test.rb [hunk 7]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
```
Comment (DMips) [inline test/mailers/organisers_mailer_test.rb]: please delete this file

File: test/mailers/previews/organisers_mailer_preview.rb [hunk 8]
```diff
@@ -0,0 +1,4 @@
+# Preview all emails at http://localhost:3000/rails/mailers/organisers_mailer
```
Comment (DMips) [inline test/mailers/previews/organisers_mailer_preview.rb]: please delete this file

File: app/views/organisers/create.html.slim (hunk 3 above)
Comment (pg20) [inline app/views/organisers/create.html.slim]: I need this because when the case fails and i do render index, it gives me error empty:nil class when is display list of organiser again. Ill look into the error now.

File: app/views/organisers/index.html.slim (hunk 6 above)
Comment (pg20) [inline app/views/organisers/index.html.slim]: Ill validate once i write them :D I have completed first the main work. Now I will do this.

Comment (pg20) [inline app/views/organisers/index.html.slim]: Also i have created a organiser_token field in user. But i think we need an extra filed like organiser_token_created_at. And then apply a validation in User Model. delete it if token is 30days old. As we can have long list of organisers, with just token. 
Also, there has to be pending_approval which for all those who have yet not filled the form or clicked the link.

File: app/models/user.rb [hunk 9]
```diff
@@ -8,4 +9,11 @@ class User < ActiveRecord::Base
   validates :email, uniqueness: true
```
Comment (DMips) [inline app/models/user.rb]: please add validation for email format also

File: app/models/user.rb [hunk 10]
```diff
@@ -1,5 +1,6 @@
 class User < ActiveRecord::Base
   authenticates_with_sorcery!
+  before_create :create_organiser_token
```
Comment (DMips) [inline app/models/user.rb]: only add this token if user is organizer

File: app/views/organisers/create.html.slim (hunk 3 above)
Comment (DMips) [inline app/views/organisers/create.html.slim]: and we definitely need to delete this one ;) if you need please call `@organisers = User.where(role: 1)` in some else

File: config/routes.rb [hunk 11]
```diff
@@ -2,8 +2,7 @@
 
 
   root 'home#index'
-
-  resources :organisers, only: [:index]
+  resources :organisers # , only: [:index, :create, :update, :edit]
```
Comment (DMips) [inline config/routes.rb]: please delete this commented part

File: app/controllers/organisers_controller.rb [hunk 12]
```diff
@@ -4,10 +4,55 @@ class OrganisersController < ApplicationController
   def index
     @organisers = User.where(role: 1)
   end
+  
+  def create
+    @user = User.find_by(email: params[:email])
+    if @user.blank?
```
Comment (DMips) [inline app/controllers/organisers_controller.rb]: if you add proper validation we won't need this if `if @user.blank?` but just `if @user.save`

File: app/models/user.rb (hunk 9 above)
Comment (DMips) [inline app/models/user.rb]: @pg20 about uniqueness - please add `uniqueness: { case_sensitive: false }` so our app will treat "u@email.com" and "U@EMAIL.COM" as same string :)
and I see you added index on email  - that together with this improved validation in modelwill be enough to ensure we don't have 2 same users in our datatabse :)

File: app/models/user.rb (hunk 10 above)
Comment (pg20) [inline app/models/user.rb]: the check cant be possible as, new user is created with a token, so we save the new user as organiser in update fn after filling the form.

File: app/views/organisers/index.html.slim (hunk 6 above)
Comment (pg20) [inline app/views/organisers/index.html.slim]: How to display error messages using form_tag? I am getting a way to do it using for_for which have user object attached to it. But I have no clue How to send validation error messages to view using form_tag? Also How you display error messages ? Using :message in validates, or making another fn ?

File: app/models/user.rb (hunk 10 above)
Comment (DMips) [inline app/models/user.rb]: just add `hidden_field_tag` for the role when adding new email - that way you'll create user with role organizer so you can make such check ;)

File: app/views/organisers/index.html.slim (hunk 6 above)
Comment (DMips) [inline app/views/organisers/index.html.slim]: because our form contains only one input we can use flash for that so in this else https://github.com/LearnITGirl/WebPlatform/pull/14/files#diff-11b31c89a012a8032400be786ecb15e5R15 instead of assigning "User exist" to flash please do this `@user.errors.full_messages.join(", ")` - that will shows us all errors

File: app/controllers/organisers_controller.rb [hunk 13]
```diff
@@ -2,12 +2,58 @@ class OrganisersController < ApplicationController
    before_action :require_organiser
 
   def index
-    @organisers = User.where(role: 1)
+    display_organisers
+  end
+  
+  def create
+    @user = User.find_by(email: params[:email])
```
Comment (DMips) [inline app/controllers/organisers_controller.rb]: why are you using `find_by` if we're creating new user here so obviously it won't exist in database yet?

File: app/controllers/organisers_controller.rb [hunk 14]
```diff
@@ -2,12 +2,58 @@ class OrganisersController < ApplicationController
...
+       redirect_to organisers_path, notice: 'Instructions have been sent to the email'
+    else
+       flash.now[:alert] = "User exist"
+       display_organisers
+       render "index"
+
+    end
+  end
+
+  def edit
```
Comment (DMips) [inline app/controllers/organisers_controller.rb]: can you move `edit` and `update` to another controller like organiser_registration or something? later we want to use `edit` and `update` in this controller for editing organiser which was already created

File: app/controllers/organisers_controller.rb (hunk 13 above)
Comment (DMips) [inline app/controllers/organisers_controller.rb]: just use `@user = User.new(email: params[:email])`

File: app/models/user.rb (hunk 10 above)
Comment (pg20) [inline app/models/user.rb]: Saving role as 1, even before organiser fills the form, the new organiser is displayed in list of added organisers. So is it required or should i leave ? Also how to diffrentiate added and pending organiser ? Next task is asana is  to display list of pending organisers.

Comment (DMips) [inline app/models/user.rb]: pending organizer will be the one that still has organiser_token - the one who clicked on the link won't have it anymore as you're clearing this field when confirming organizer right?

And I'm not sure I'm following question about required thing ;)

File: app/controllers/organisers_registration_controller.rb [hunk 15]
```diff
@@ -0,0 +1,55 @@
+class OrganisersRegistrationController < ApplicationController
+ before_action :require_organiser
+ def create
+    @user = User.new(email: params[:email])
```
Comment (DMips) [inline app/controllers/organisers_registration_controller.rb]: right sorry you can remove hidden_field form organiser form and just call `@user = User.new(email: params[:email], role: 1)` as we want to call this method only when creating organiser so we can do that :) And obviously current line 5 should be removed.

Btw why did you move create to this controller?

File: app/models/user.rb [hunk 16]
```diff
@@ -1,11 +1,19 @@
...
 
   validates :password, length: { minimum: 6 }, if: -> { password || password_confirmation }
   validates :password, confirmation: true, if: -> { password || password_confirmation }
   validates :password_confirmation, presence: true, if: -> { password || password_confirmation }
 
-  validates :email, uniqueness: true
+  validates :email,  uniqueness: { case_sensitive: false } , presence: true,  format: { with: REGEXP_EMAIL }, on: :create
```
Comment (DMips) [inline app/models/user.rb]: we always want to validate email (even when editing) so let's remove `on: "create`

File: app/models/user.rb [hunk 17]
```diff
@@ -1,11 +1,19 @@
...
 
   enum role: {organizer: 1, mentee: 2, mentor: 3}
+
+  private
+    def create_organiser_token
+      if self.role == 'organizer'
+         self.organiser_token = SecureRandom.hex(8)
```
Comment (DMips) [inline app/models/user.rb]: you can write here `self.organiser_token = SecureRandom.hex(8) if role == 'organizer'` so that will be 1 line instead of 3 :)

File: app/controllers/organisers_registration_controller.rb [hunk 18]
```diff
@@ -0,0 +1,55 @@
+class OrganisersRegistrationController < ApplicationController
```
Comment (DMips) [inline app/controllers/organisers_registration_controller.rb]: it should be `OrganiserRegistrationsController` plural form :)

File: app/models/user.rb (hunk 10 above)
Comment (pg20) [inline app/models/user.rb]: Hi i havent destroyed the token in update action. What should i do ? change user.organiser_token = null ?

File: app/models/user.rb (hunk 16 above)
Comment (pg20) [inline app/models/user.rb]: There is no validation on email in edit form as email field is automatically taken as input and we dont ask from the user.

File: app/models/user.rb (hunk 10 above)
Comment (DMips) [inline app/models/user.rb]: 1. yes if user is successfully updates please set this token to nil
2. you need to change this query instead of `@organisers = User.where(role: 1)` you need to use `@organisers = User.where(role: 1, organiser_token: nil)` for confirmed organisers and `@organisers = User.where(role: 1).where.not(organiser_token: nil)` for pending ones

File: app/controllers/organiser_registrations_controller.rb [hunk 19]
```diff
@@ -0,0 +1,37 @@
+class OrganiserRegistrationsController < ApplicationController
+ before_action :require_organiser
```
Comment (DMips) [inline app/controllers/organiser_registrations_controller.rb]: when you click on this link from your inbox you won't be logged in to the system so we can't require that user will be logged it (he can't do that yet - that's why we're showing this page ;)) so please remove this `before_filter`

**Comment by DMips** [comment]
looks good :) please just merge master to this branch and it will be ready :)

on branch master please do:
git pull --rebase
then
git checkout add-organiser
git merge master
git push

if you have any problems let me know and we can do it together or something ;)
